packages/flutter_tools/lib/src/xcode_project.dart

```diff
@@ -264,9 +280,9 @@ class IosProject extends XcodeBasedProject {
       }
     }
     if (productName == null) {
-      globals.printTrace('FULL_PRODUCT_NAME not present, defaulting to ${XcodeBasedProject._hostAppProjectName}');
+      globals.printTrace('FULL_PRODUCT_NAME not present, defaulting to $hostAppProjectName');
     }
-    return productName ?? '${XcodeBasedProject._hostAppProjectName}.app';
+    return productName ?? '$hostAppProjectName.app';
```

Comment:
I changed my project and workspace names and it still builds Runner.app.

This wouldn't be changed by the Xcode project name, the intent here was to fall back to `Runner.app` which would be the default from the `flutter create` template:

https://github.com/flutter/flutter/blob/635337ca1556b7a6cb50bb9ca0d36721e9772234/packages/flutter_tools/templates/app_shared/ios-objc.tmpl/Runner.xcodeproj/project.pbxproj.tmpl#L390

(`TARGET_NAME` is Runner)

<img width="161" alt="Screenshot 2023-05-01 at 4 29 26 PM" src="https://user-images.githubusercontent.com/682784/235548800-3d45d89e-c64d-4397-9ec3-08cb4397010a.png">

So this should be treated more like the directory name.  Though at that point `_defaultHostAppProjectDirectoryName` isn't a good name since it's representing the fallback for project name, directory name, and product name.